File: ts/openebx-lib/test/block-builder.test.ts
import { describe, expect, test, beforeEach, it } from '@jest/globals'
import BlockHeader from '../src/block-header'
import Block from '../src/block'
import Tx from '../src/tx'
import BufferWriter from '../src/buffer-writer'
import BufferReader from '../src/buffer-reader'
import BlockBuilder from '../src/block-builder'
import Script from '../src/script'

describe('BlockBuilder', () => {
  test('fromBlock', () => {
    const bh = new BlockHeader(
      1,
      new Uint8Array(32),
      new Uint8Array(32),
      0n,
      new Uint8Array(32),
      new Uint8Array(32),
      0n,
    )
    const tx = new Tx(1, [], [], 0n)
    const block = new Block(bh, [tx])
    const bb = BlockBuilder.fromBlock(block)
    expect(bb.header.version).toBe(bh.version)
    expect(bb.header.prevBlockId).toEqual(bh.prevBlockId)
    expect(bb.header.merkleRoot).toEqual(bh.merkleRoot)
    expect(bb.header.timestamp).toBe(bh.timestamp)
    expect(bb.header.target).toEqual(bh.target)
  })

  test('fromGenesis', () => {
    const target = new Uint8Array(32)
    const outputScript = new Script()
    const outputAmount = 0n
    const bb = BlockBuilder.fromGenesis(target, outputScript, outputAmount)
    expect(bb.header.version).toBe(1)
    expect(bb.header.prevBlockId).toEqual(new Uint8Array(32))
    expect(bb.header.merkleRoot).toEqual(bb.merkleTxs.root)
    expect(bb.header.timestamp).toBeLessThanOrEqual(new Date().getTime() / 1000)
    expect(bb.header.target).toEqual(target)
  })

  test('fromPrevBlockHeader', () => {
    const outputScript = new Script()
    const outputAmount = 0n
    const target = new Uint8Array(32)
    const prevBlockHeader = new BlockHeader(
      1,
      new Uint8Array(32),
      new Uint8Array(32),
      0n,
      target,
      new Uint8Array(32),
      0n,
    )
    const bb = BlockBuilder.fromPrevBlockHeader(
      prevBlockHeader,
      null,
      outputScript,
      outputAmount,
    )
    expect(bb.header.version).toBe(1)
    expect(bb.header.prevBlockId).toEqual(prevBlockHeader.id())
    expect(bb.header.merkleRoot).toEqual(bb.merkleTxs.root)
    expect(bb.header.timestamp).toBeLessThanOrEqual(new Date().getTime() / 1000)
    expect(bb.header.target).toEqual(target)
  })

  test('toBlock', () => {
    const bh = new BlockHeader(
      1,
      new Uint8Array(32),
      new Uint8Array(32),
      0n,
      new Uint8Array(32),
      new Uint8Array(32),
      0n,
    )
    const tx = new Tx(1, [], [], 0n)
    const block = new Block(bh, [tx])
    const bb = BlockBuilder.fromBlock(block)
    const block2 = bb.toBlock()
    expect(block2.header.version).toBe(bh.version)
    expect(block2.header.prevBlockId).toEqual(bh.prevBlockId)
    expect(block2.header.merkleRoot).toEqual(bh.merkleRoot)
    expect(bb.header.timestamp).toEqual(0n)
    expect(block2.header.target).toEqual(bh.target)
  })
})

File: rs/earthbucks_lib/tests/block_builder_test.rs
use earthbucks_lib::block::Block;
use earthbucks_lib::block_builder::BlockBuilder;
use earthbucks_lib::header::Header;
use earthbucks_lib::tx::Tx;

#[test]
fn test_from_block() {
    let bh = Header {
        version: 1,
        prev_block_id: [0; 32],
        merkle_root: [0; 32],
        timestamp: 0,
        block_num: 0,
        target: [0; 32],
        nonce: [0; 32],
        work_ser_algo: 0,
        work_ser_hash: [0; 32],
        work_par_algo: 0,
        work_par_hash: [0; 32],
    };
    let tx = Tx::new(1, vec![], vec![], 0);
    let block = Block::new(bh.clone(), vec![tx]);
    let bb = BlockBuilder::from_block(block);
    assert_eq!(bb.header.version, bh.version);
    assert_eq!(bb.header.prev_block_id, bh.prev_block_id);
    assert_eq!(bb.header.merkle_root, bh.merkle_root);
    assert_eq!(bb.header.timestamp, bh.timestamp);
    assert_eq!(bb.header.target, bh.target);
}
